WordPress/Classes/ViewRelated/Menus/Views/MenuItemSourceView.m

```diff
@@ -0,0 +1,475 @@
+}
+
+- (NSFetchRequest *)fetchRequest
+{
+    // overrided in subclasses
+    return nil;
+}
+
+- (NSPredicate *)defaultFetchRequestPredicate
+{
+    // overrided in subclasses if needed
+    NSPredicate *predicate = [NSPredicate predicateWithFormat:@"blog == %@", [self blog]];
+    return predicate;
+}
+
+- (NSString *)fetechedResultsControllerSectionNameKeyPath
+{
+    // overrided in subclasses
+    return nil;
+}
+
+- (void)performResultsControllerFetchRequest
+{
+    if (!self.resultsController) {
+        return;
+    }
+    NSError *error;
+    if (![self.resultsController performFetch:&error]) {
+        DDLogError(@"Error occurred preforming fetch for MenuItem source of type: %@ error: %@", [self sourceItemType], error);
+    }
+    [self toggleNoResultsIndiciator];
+}
+
+- (BOOL)fetchedResultsAreEmpty
+{
+    return self.resultsController.fetchedObjects.count == 0;
+}
+
+- (void)deselectVisibleSourceCellsIfNeeded
+{
+    NSArray *cells = [self.tableView visibleCells];
+    for (MenuItemSourceCell *cell in cells) {
+        if (![cell isKindOfClass:[MenuItemSourceCell class]]) {
+            continue;
+        }
+        if (cell.sourceSelected) {
+            cell.sourceSelected = NO;
+        }
+    }
+}
+
+#pragma mark - subclass configuration
+
+- (void)configureSourceCell:(MenuItemSourceCell *)cell forIndexPath:(NSIndexPath *)indexPath
+{
+    // overrided in subclasses
+}
+
+- (void)scrollingWillDisplayEndOfTableView:(UITableView *)tableView
+{
+    // overrided in subclasses
+}
+
+#pragma mark - UIScrollView
+
+- (void)scrollViewWillBeginDragging:(UIScrollView *)scrollView
+{
+    self.observeUserScrollingForEndOfTableView = YES;
+}
+
+- (void)scrollViewDidEndDragging:(UIScrollView *)scrollView willDecelerate:(BOOL)decelerate
+{
+    if (!decelerate) {
+        self.observeUserScrollingForEndOfTableView = NO;
+    }
+}
+
+- (void)scrollViewDidEndDecelerating:(UIScrollView *)scrollView
+{
+    self.observeUserScrollingForEndOfTableView = NO;
+}
+
+#pragma mark - UITableView
+
+- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView
+{
+    return self.resultsController.sections.count;
+}
+
+- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section
+{
+    id <NSFetchedResultsSectionInfo> sectionInfo = [[self.resultsController sections] objectAtIndex:section];
+    return [sectionInfo numberOfObjects];
+}
+
+- (CGFloat)tableView:(UITableView *)tableView estimatedHeightForRowAtIndexPath:(NSIndexPath *)indexPath
+{
+    return 50;
+}
+
+- (CGFloat)tableView:(UITableView *)tableView heightForRowAtIndexPath:(NSIndexPath *)indexPath
+{
+    return UITableViewAutomaticDimension;
+}
+
+- (void)tableView:(UITableView *)tableView willDisplayCell:(nonnull UITableViewCell *)cell forRowAtIndexPath:(nonnull NSIndexPath *)indexPath
+{
+    NSInteger numSections = self.resultsController.sections.count;
+    if (indexPath.section == numSections - 1 && self.observeUserScrollingForEndOfTableView) {
```

Comment:
It would be nice if we could return early to avoid some of the nested conditionals.